### PR TITLE
Fix spigot whitelist command on newer Paper versions

### LINKED ISSUE #444: Error when trying to whitelist bedrock player
whenever I try to run "fwhitelist add [bedrock username]",  I get this error in the console:

[https://pastebin.com/xJ4KhZCb]

checked with multiple people and I got the same error
Thanks for help in advance!

### PR DESCRIPTION
**Pull request by HideyBoi**
I compiled and used this fix myself for a while and it is currently working. It let me add and remove players from the whitelist without any errors. 

I am no expert on Java code, unsure how elegant of a fix this is, but it works for now.

Closes #444

### CODE REVIEW

**Review by Kas-tle — CHANGES_REQUESTED** [review]
Instead of calling the scheduler directly could you please use SpigotVersionSpecificMethods#maybeSchedule via SpigotCommandUtil?

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 1]
```diff
@@ -25,10 +25,17 @@
...
 
+import static org.bukkit.Bukkit.getServer;
+
 import com.mojang.authlib.GameProfile;
 import java.util.UUID;
 import org.bukkit.Bukkit;
 import org.bukkit.OfflinePlayer;
+import org.geysermc.floodgate.FloodgatePlatform;
+import org.geysermc.floodgate.SpigotPlugin;
+import org.geysermc.floodgate.api.FloodgateApi;
```
Comment (onebeastchris) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:36]: Is this (and the other new imports) needed?
Doesn't look like they're used

**Review by onebeastchris — CHANGES_REQUESTED** [review]

**Comment by HideyBoi** [comment]
@Kas-tle do you mean like this?

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java (hunk 1 above)
Comment (HideyBoi) [reply spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java]: You're right. That was accidentally left over from when I was trying to figure out a solution. I've removed them.

File: spigot/src/main/java/org/geysermc/floodgate/util/SpigotCommandUtil.java [hunk 2]
```diff
@@ -122,11 +123,23 @@ public void kickPlayer(Object player, String message) {
...
+    public void setWhitelist(Object offlinePlayer, Boolean isWhitelisted) {
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                OfflinePlayer player = (OfflinePlayer)offlinePlayer;
+                player.setWhitelisted(isWhitelisted);
+            }
+        };
+
+        versionSpecificMethods.maybeSchedule(runnable);
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/SpigotCommandUtil.java:143]: Move this to `WhitelistUtils`. And the `isWhitelisted` parameter can be a primitive.

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 3]
```diff
@@ -25,54 +25,64 @@
 
 package org.geysermc.floodgate.util;
 
+import static org.bukkit.Bukkit.getServer;
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:28]: Static import isn't necessary here

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 4]
```diff
@@ -25,54 +25,64 @@
...
+
 import com.mojang.authlib.GameProfile;
 import java.util.UUID;
-import org.bukkit.Bukkit;
 import org.bukkit.OfflinePlayer;
 
+
+
 @SuppressWarnings("ConstantConditions")
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:36]: Undo this diff

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 5]
```diff
@@ -25,54 +25,64 @@
...
 public final class WhitelistUtils {
 
     /**
      * Whitelist the given Bedrock player.
      *
-     * @param uuid     the UUID of the Bedrock player to be whitelisted
-     * @param username the username of the Bedrock player to be whitelisted
+     * @param uuid              the UUID of the Bedrock player to be whitelisted
+     * @param username          the username of the Bedrock player to be whitelisted
+     * @param spigotCommandUtil
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:44]: Undo the diff for the first two lines, finish javadoc for the new parameter

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 6]
```diff
@@ -25,54 +25,64 @@
...
-                Bukkit.getServer(), profile
+                getServer(), profile
         );
         if (player.isWhitelisted()) {
             return false;
         }
-        player.setWhitelisted(true);
+
+        spigotCommandUtil.setWhitelist(player, true);
+
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:59]: Remove the two spaces for less diff

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 7]
```diff
@@ -25,54 +25,64 @@
...
     }
 
     /**
      * Removes the given Bedrock player from the whitelist.
      *
-     * @param uuid     the UUID of the Bedrock player to be removed
-     * @param username the username of the Bedrock player to be removed
+     * @param uuid              the UUID of the Bedrock player to be removed
+     * @param username          the username of the Bedrock player to be removed
+     * @param spigotCommandUtil
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:68]: same thing here as above

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 8]
```diff
@@ -25,54 +25,64 @@
...
      * @return true if the player has been removed from the whitelist, false if the player wasn't
      * whitelisted
      */
-    public static boolean removePlayer(UUID uuid, String username) {
+    public static boolean removePlayer(UUID uuid, String username,
+                                       SpigotCommandUtil spigotCommandUtil) {
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:73]: Single line

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 9]
```diff
@@ -25,54 +25,64 @@
...
-                Bukkit.getServer(), profile
+                getServer(), profile
         );
         if (!player.isWhitelisted()) {
             return false;
         }
-        player.setWhitelisted(false);
+
+        spigotCommandUtil.setWhitelist(player, false);
+
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:85]: Remove the 2 spaces

**Review by Konicai — CHANGES_REQUESTED** [review]

File: spigot/src/main/java/org/geysermc/floodgate/util/SpigotCommandUtil.java (hunk 2 above)
Comment (Konicai) [reply spigot/src/main/java/org/geysermc/floodgate/util/SpigotCommandUtil.java:143]: Also, the player parameter can and should be type of `OfflinePlayer`. And rename `isWhitelisted` to `whitelist`

**Review by Kas-tle — APPROVED** [review]

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 10]
```diff
@@ -72,7 +74,18 @@ public static boolean removePlayer(UUID uuid, String username) {
...
+    static void setWhitelist(OfflinePlayer player, boolean whitelist, SpigotVersionSpecificMethods versionSpecificMethods) {
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                player.setWhitelisted(whitelist);
+            }
+        };
+
+        versionSpecificMethods.maybeSchedule(runnable);
+    }
```
Comment (Tim203) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:90]:
```suggestion
    static void setWhitelisted(OfflinePlayer player, boolean whitelisted, SpigotVersionSpecificMethods versionSpecificMethods) {
        versionSpecificMethods.maybeSchedule(() -> player.setWhitelisted(whitelisted));
    }

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 11]
```diff
@@ -25,9 +25,9 @@
 
 package org.geysermc.floodgate.util;
 
+import org.bukkit.Bukkit;
 import com.mojang.authlib.GameProfile;
 import java.util.UUID;
-import org.bukkit.Bukkit;
 import org.bukkit.OfflinePlayer;
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:31]: Revert this diff please

File: spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java [hunk 12]
```diff
@@ -72,7 +74,13 @@ public static boolean removePlayer(UUID uuid, String username) {
...
         }
-        player.setWhitelisted(false);
+        setWhitelist(player, false, versionSpecificMethods);
         return true;
     }
+
+    static void setWhitelist(OfflinePlayer player, boolean whitelist, SpigotVersionSpecificMethods versionSpecificMethods) {
+        versionSpecificMethods.maybeSchedule(() -> {
+            player.setWhitelisted(whitelist);
+        });
```
Comment (Konicai) [inline spigot/src/main/java/org/geysermc/floodgate/util/WhitelistUtils.java:84]: one line please

**Review by Konicai — APPROVED** [review]

**Review by onebeastchris — APPROVED** [review]

**Pull request merged**